File: app/src/main/java/com/fusoft/walkboner/CreatePostActivity.java
package com.fusoft.walkboner;

import android.animation.LayoutTransition;
import android.app.Activity;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.fusoft.walkboner.auth.Authentication;
import com.fusoft.walkboner.auth.AuthenticationListener;
import com.fusoft.walkboner.auth.UserInfoListener;
import com.fusoft.walkboner.database.DeleteImage;
import com.fusoft.walkboner.database.ImageUploadListener;
import com.fusoft.walkboner.database.StorageDirectory;
import com.fusoft.walkboner.database.UploadImage;
import com.fusoft.walkboner.models.User;
import com.fusoft.walkboner.utils.UidGenerator;
import com.fusoft.walkboner.views.dialogs.ErrorDialog;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textview.MaterialTextView;
import com.google.firebase.firestore.FirebaseFirestore;

import java.sql.Timestamp;
import java.util.HashMap;

import de.dlyt.yanndroid.oneui.dialog.ProgressDialog;
import de.dlyt.yanndroid.oneui.view.Toast;
import de.dlyt.yanndroid.oneui.widget.ProgressBar;
import de.dlyt.yanndroid.oneui.widget.Spinner;
import de.dlyt.yanndroid.oneui.widget.Switch;

public class CreatePostActivity extends AppCompatActivity {
    private EditText descriptionEdt;
    private ImageView image;
    private Spinner celebritiesSpinner;
    private Switch allowCommentsSwitch;
    private MaterialButton addButton;
    private MaterialTextView maxTextSize;
    private ProgressDialog loading;
    private LinearLayout mainLinear;
    private LinearLayout uploadingImageLinear;
    private ProgressBar imageUploadingProgressBar;
    private MaterialTextView maxLettersText;

    private Uri imageStoragePath;
    private String imageUrl = "";
    private String documentUid = "";

    private boolean allowComments = true;

    FirebaseFirestore firestore;

    private Authentication authentication;
    private User user;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_post);

        loading = new ProgressDialog(CreatePostActivity.this);
        loading.setIndeterminate(true);
        loading.setCancelable(false);
        loading.setCanceledOnTouchOutside(false);
        loading.setProgressStyle(ProgressDialog.STYLE_CIRCLE);

        firestore = FirebaseFirestore.getInstance();

        authentication = new Authentication(null);

        authentication.getUserData(new UserInfoListener() {
            @Override
            public void OnUserDataReceived(User userD) {
                user = userD;
                addButton.setEnabled(true);
            }

            @Override
            public void OnUserNotFinded() {

            }

            @Override
            public void OnError(String reason) {

            }
        });

        initView();
        setup();
    }

    private void initView() {
        descriptionEdt = (EditText) findViewById(R.id.description_edt);
        image = (ImageView) findViewById(R.id.image);
        celebritiesSpinner = (Spinner) findViewById(R.id.celebrities_spinner);
        allowCommentsSwitch = (Switch) findViewById(R.id.allow_comments_switch);
        addButton = (MaterialButton) findViewById(R.id.add_button);
        maxTextSize = findViewById(R.id.max_letters_text);
        mainLinear = (LinearLayout) findViewById(R.id.main_linear);
        mainLinear.getLayoutTransition().enableTransitionType(LayoutTransition.CHANGING);
        uploadingImageLinear = (LinearLayout) findViewById(R.id.uploading_image_linear);
        imageUploadingProgressBar = (ProgressBar) findViewById(R.id.image_uploading_progress_bar);
        maxLettersText = (MaterialTextView) findViewById(R.id.max_letters_text);
    }

    private void setup() {
        image.setOnClickListener(v -> {
            imageChooser();
        });

        addButton.setOnClickListener(v -> {
            if (imageStoragePath != null) {
                if (!imageUrl.isEmpty()) {
                    DeleteImage.Delete(CreatePostActivity.this, imageUrl, new DeleteImage.DeleteTask() {
                        @Override
                        public void OnDeleted() {

                        }

                        @Override
                        public void OnError(String reason) {

                        }
                    });
                }
                uploadingImageLinear.setVisibility(View.VISIBLE);
                new UploadImage(CreatePostActivity.this, StorageDirectory.POSTS_PATH, imageStoragePath, new ImageUploadListener() {
                    @Override
                    public void OnImageUploaded(String imageUrlString) {
                        imageUrl = imageUrlString;
                        firestore.collection("posts").add(defaultMap()).addOnSuccessListener(documentReference -> {
                            HashMap<String, Object> map = new HashMap<>();
                            map.put("documentUid", documentReference.getId());
                            firestore.collection("posts").document(documentReference.getId()).update(map);
                            finish();
                        }).addOnFailureListener(e -> {
                            new ErrorDialog().SimpleErrorDialog(CreatePostActivity.this, e.getMessage());
                        });
                    }

                    @Override
                    public void Progress(int value) {
                        imageUploadingProgressBar.setProgress(value);
                    }

                    @Override
                    public void OnError(String reason) {
                        Toast.makeText(CreatePostActivity.this, reason, Toast.LENGTH_SHORT).show();
                    }
                });
            } else {
                Toast.makeText(CreatePostActivity.this, "Wybierz Zdjęcie, które chcesz załączyć do postu", Toast.LENGTH_SHORT).show();
            }
        });

        allowCommentsSwitch.setOnCheckedChangeListener((compoundButton, b) -> {
            allowComments = b;
        });

        descriptionEdt.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                if (editable.length() >= 201) {
                    maxTextSize.setTextColor(getColor(android.R.color.holo_red_light));
                } else {
                    maxTextSize.setTextColor(getColor(de.dlyt.yanndroid.oneui.R.color.sesl_secondary_text));
                }
            }
        });
    }

    private void imageChooser() {
        Intent i = new Intent();
        i.setType("image/*");
        i.setAction(Intent.ACTION_GET_CONTENT);

        activityResultLauncher.launch(i);
    }

    ActivityResultLauncher<Intent> activityResultLauncher = registerForActivityResult(
            new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == Activity.RESULT_OK) {
                    imageStoragePath = result.getData().getData();
                }
            });

    @NonNull
    private HashMap<String, Object> defaultMap() {
        Timestamp timestamp = new Timestamp(System.currentTimeMillis());

        HashMap<String, Object> map = new HashMap<>();
        map.put("userUid", user.getUserUid());
        map.put("postUid", UidGenerator.Generate());
        map.put("postDescription", descriptionEdt.getText().toString());
        map.put("postImage", imageUrl);
        map.put("documentUid", "");
        map.put("createdAt", String.valueOf(timestamp.getTime()));
        map.put("showsCelebrity", "celebrityUid");
        map.put("allowComments", String.valueOf(allowComments));

        return map;
    }
}

File: app/src/main/java/com/fusoft/walkboner/CreatePostActivity.kt
package com.fusoft.walkboner

import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.materialswitch.MaterialSwitch
import com.google.android.material.button.MaterialButton
import com.google.android.material.textview.MaterialTextView
import com.fusoft.walkboner.views.LoadingView
import com.google.android.material.progressindicator.LinearProgressIndicator
import com.google.firebase.firestore.FirebaseFirestore
import java.lang.Boolean
import android.os.Bundle
import com.fusoft.walkboner.R
import com.fusoft.walkboner.auth.UserInfoListener
import android.animation.LayoutTransition
import com.fusoft.walkboner.database.DeleteImage
import com.fusoft.walkboner.database.DeleteImage.DeleteTask
import com.fusoft.walkboner.database.UploadImage
import com.fusoft.walkboner.database.StorageDirectory
import com.fusoft.walkboner.database.ImageUploadListener
import com.google.android.gms.tasks.OnSuccessListener
import com.google.android.gms.tasks.OnFailureListener
import com.fusoft.walkboner.views.dialogs.ErrorDialog
import android.app.Activity
import android.content.Intent
import android.net.Uri
import android.view.View
import android.widget.*
import androidx.activity.ComponentActivity
import androidx.activity.result.ActivityResult
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContract
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import com.fusoft.walkboner.adapters.spinner.InfluencersNicksAdapter
import com.fusoft.walkboner.auth.Authentication
import com.fusoft.walkboner.databinding.ActivityCreatePostBinding
import com.fusoft.walkboner.models.User
import com.fusoft.walkboner.singletons.UserSingletone
import com.squareup.picasso.Picasso
import com.fusoft.walkboner.utils.UidGenerator
import com.google.firebase.firestore.DocumentReference
import com.google.firebase.firestore.QuerySnapshot
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.lang.Exception
import java.sql.Timestamp
import java.util.HashMap
import kotlin.coroutines.CoroutineContext

class CreatePostActivity : AppCompatActivity(), CoroutineScope {
    private var job: Job = Job()

    override val coroutineContext: CoroutineContext
        get() = Dispatchers.Main + job

    /* ------------------------------------------ */

    lateinit var binding: ActivityCreatePostBinding

    private var imageStoragePath: Uri? = null
    private var imageUrl = ""
    private var allowComments = true

    var firestore: FirebaseFirestore? = null
    private var authentication: Authentication? = null
    private var user: User? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityCreatePostBinding.inflate(layoutInflater)
        setContentView(binding.root)
        firestore = FirebaseFirestore.getInstance()

        user = UserSingletone.getInstance().user
        binding.addButton.isEnabled = true

        initView()
        setup()
    }

    private fun initView() {
        binding.mainLinear.layoutTransition.enableTransitionType(LayoutTransition.CHANGING)

        launch { loadInfluencersDataToSpinner() }
    }

    private fun setup() {
        binding.image.setOnClickListener { v: View? -> imageChooser() }
        binding.addButton.setOnClickListener { v: View? ->
            if (imageStoragePath != null) {
                if (!imageUrl.isEmpty()) {
                    DeleteImage.Delete(this@CreatePostActivity, imageUrl, object : DeleteTask {
                        override fun OnDeleted() {}
                        override fun OnError(reason: String) {}
                    })
                }

                binding.uploadingImageLinear.visibility = View.VISIBLE
                UploadImage(
                    this@CreatePostActivity,
                    StorageDirectory.POSTS_PATH,
                    imageStoragePath,
                    object : ImageUploadListener {
                        override fun OnImageUploaded(imageUrlString: String) {
                            imageUrl = imageUrlString
                            firestore!!.collection("posts").add(defaultMap())
                                .addOnSuccessListener { documentReference: DocumentReference ->
                                    val map = HashMap<String, Any>()
                                    map["documentUid"] = documentReference.id
                                    firestore!!.collection("posts").document(documentReference.id)
                                        .update(map)
                                    finish()
                                }.addOnFailureListener { e: Exception ->
                                ErrorDialog().SimpleErrorDialog(
                                    this@CreatePostActivity,
                                    e.message
                                )
                            }
                        }

                        override fun Progress(value: Int) {
                            binding.imageUploadingProgressBar.progress = value
                        }

                        override fun OnError(reason: String) {
                            Toast.makeText(this@CreatePostActivity, reason, Toast.LENGTH_SHORT)
                                .show()
                        }
                    })
            } else {
                Toast.makeText(
                    this@CreatePostActivity,
                    "Wybierz Zdjęcie, które chcesz załączyć do postu",
                    Toast.LENGTH_SHORT
                ).show()
            }
        }

        binding.allowCommentsSwitch.setOnCheckedChangeListener { buttonView, isChecked ->
            allowComments = isChecked
        }
    }

    private suspend fun loadInfluencersDataToSpinner() {
        var influencers: QuerySnapshot? = null
        var influencersData: ArrayList<HashMap<String, String>>? = arrayListOf()
        var spinnerAdapter: InfluencersNicksAdapter? = null

        firestore?.collection("influencers")?.get()?.addOnSuccessListener { querySnapshot ->
            influencers = querySnapshot
        }?.await()

        if (influencers == null || influencers?.isEmpty == true) {
            Toast.makeText(this@CreatePostActivity, "Nie udało się załatować listy influencerek!", Toast.LENGTH_SHORT).show()
            return
        }

        for (influencer in influencers?.documents!!) {
            if (!Boolean.parseBoolean(influencer?.get("isHidden").toString())) {
                val influ = HashMap<String, String>()
                influ.put("nickname", influencer?.getString("influencerNickName").toString())
                influ.put("uid", influencer?.id.toString())
                influencersData?.add(influ)
            }
        }

        // sort list a-z
        influencersData?.sortBy { it.get("nickname").toString() }

        var non = HashMap<String, String>()
        non.put("nickname", "Nikt")
        non.put("uid", "null")

        // FIXME wtf
        influencersData?.reverse()
        influencersData?.add(non)
        influencersData?.reverse()

        spinnerAdapter = InfluencersNicksAdapter(this@CreatePostActivity, influencersData)
        binding.celebritiesSpinner.adapter = spinnerAdapter
    }

    private fun imageChooser() {
        val i = Intent()
        i.type = "image/*"
        i.action = Intent.ACTION_GET_CONTENT
        activityResultLauncher.launch(i)
    }

    /*
     * FIXME: Wait For Fix from Google
     * Because of Android Studio Bug
     * we need to add this extension to dismiss
     * registerForActivityResult unresolved reference
     * https://stackoverflow.com/a/74139700/13680873
     */
    fun <I, O> Activity.registerForActivityResult(
        contract: ActivityResultContract<I, O>,
        callback: ActivityResultCallback<O>
    ) = (this as ComponentActivity).registerForActivityResult(contract, callback)
    /* This Function does nothing */

    var activityResultLauncher: ActivityResultLauncher<Intent> =
        registerForActivityResult<Intent, ActivityResult>(
            ActivityResultContracts.StartActivityForResult(),
            ActivityResultCallback { result: ActivityResult ->
                if (result.resultCode == RESULT_OK) {
                    imageStoragePath = result.data!!.data
                    Picasso.get().load(imageStoragePath).into(binding.image)
                }
            })

    private fun defaultMap(): HashMap<String, Any> {
        val timestamp = Timestamp(System.currentTimeMillis())
        val map = HashMap<String, Any>()
        val hashMap: HashMap<String, String> = binding.celebritiesSpinner.getItemAtPosition(binding.celebritiesSpinner.selectedItemPosition) as HashMap<String, String>
        map["userUid"] = user!!.userUid
        map["postUid"] = UidGenerator.Generate()
        map["postDescription"] = binding.descriptionEdt.text.toString()
        map["postImage"] = imageUrl
        map["documentUid"] = ""
        map["createdAt"] = timestamp.time.toString()
        map["showsCelebrity"] = hashMap["uid"].toString()
        map["allowComments"] = allowComments.toString()
        return map
    }
}